pkgs/build-support/binary-cache.nix

```diff
@@ -0,0 +1,77 @@
+{ stdenv, buildPackages }:
+
+# This function is for creating a flat-file binary cache, i.e. the kind created by
+# nix copy --to file:///some/path and usable as a substituter (with the file:// prefix).
+
+# For example, in the Nixpkgs repo:
+# nix-build -E 'with import ./. {}; mkBinaryCache { rootPaths = [hello]; }'
+
+{ name ? "binary-cache"
+, rootPaths
+}:
+
+stdenv.mkDerivation {
+  inherit name;
+
+  __structuredAttrs = true;
+
+  exportReferencesGraph.closure = rootPaths;
+
+  preferLocalBuild = true;
+
+  PATH = "${buildPackages.coreutils}/bin:${buildPackages.jq}/bin:${buildPackages.python3}/bin:${buildPackages.nix}/bin:${buildPackages.xz}/bin";
+
+  builder = builtins.toFile "builder" ''
+    . .attrs.sh
+
+    export out=''${outputs[out]}
+
+    mkdir $out
+    mkdir $out/nar
+
+    python <<EOF
+    import json
+    import os
+    import subprocess
+
+    with open(".attrs.json", "r") as f:
+      closures = json.load(f)["closure"]
+
+    os.chdir(os.environ["out"])
+
+    nixPrefix = "/nix/store" # TODO: general Nix prefix
+
+    with open("nix-cache-info", "w") as f:
+      f.write("StoreDir: " + nixPrefix + "\n")
+
+    def dropPrefix(path):
+      return path[len(nixPrefix + "/"):]
+
+    for item in closures:
+      narInfoHash = dropPrefix(item["path"]).split("-")[0]
+
+      xzFile = "nar/" + narInfoHash + ".nar.xz"
+      with open(xzFile, "w") as f:
+        subprocess.run("nix-store --dump %s | xz -c" % item["path"], stdout=f, shell=True)
+
+      fileHash = subprocess.run(["nix-hash", "--base32", "--type", "sha256", item["path"]], capture_output=True).stdout.decode().strip()
+      fileSize = os.path.getsize(xzFile)
+
+      # Rename the .nar.xz file to its own hash to match "nix copy" behavior
+      finalXzFile = "nar/" + fileHash + ".nar.xz"
+      os.rename(xzFile, finalXzFile)
+
+      with open(narInfoHash + ".narinfo", "w") as f:
```

Comment:
Isn't there some way to get nix to create the narinfo?

AFAICT the file format isn't specified anywhere, so if it changes this won't get updated...

Reply:
Not that I'm aware of, but I'd love to find one. It seems pretty ad-hoc, look at this CGI script in `nix-binary-cache`: https://github.com/NixOS/nixpkgs/blob/728ebb50a898afaa56c286b2b06ff93932ca9ea5/pkgs/servers/http/nix-binary-cache/nix-binary-cache.cgi.in#L63